### Excercise 1 -- Monte carlo

In [4]:
import random
import statistics
import time
import math
from multiprocessing import Pool, cpu_count

def estimate_pi_serial(num_samples):

    hits = 0

    for i in range(num_samples):
        x = random.random()
        y = random.random()

        if x * x + y * y <= 1.0:
            hits += 1

    return 4.0 * hits / num_samples



if __name__ == "__main__":
    
    NUM_SAMPLES = 10000000
    TRUE_PI = math.pi

    times = []

    for i in range (3):
        start = time.perf_counter()
        pi_estimate = estimate_pi_serial(NUM_SAMPLES)
        end = time.perf_counter()

        result = times.append(end - start)
    
    time_serial = statistics.median(times)
    error_serial = abs(pi_estimate - TRUE_PI)
    print(f"Serial Estimate: {pi_estimate:.6f}, Time: {time_serial:.4f} seconds, Error: {error_serial:.6f}")

Serial Estimate: 3.141236, Time: 2.0202 seconds, Error: 0.000357


Serial Estimate: 3.141592, Time: 1.7396 seconds, Error: 0.000001
Serial Estimate: 3.141203, Time: 1.8428 seconds, Error: 0.000390
Serial Estimate: 3.141236, Time: 2.0202 seconds, Error: 0.000357

### Questions

####  How accurate is the estimate? Run several times — does it vary?
- The estimate is accurate to the 4th decimal.
- Yes, everything after the 4th decimal changes with each run.

####  What is the serial time? This will be your speedup denominator in E3.
- Median time: Time: 1.7396